File: threading/cpp/18_traffic_light.cpp
/*
 * QUESTION 18: Traffic Light Controlled Intersection (LeetCode 1279)
 * ==================================================================
 * 
 * Problem: Cars from two roads approach intersection. Control traffic light
 * to ensure cars from only one road pass at a time.
 * 
 * Key Concepts: Mutual exclusion, state machines, thread coordination
 * 
 * Compile: g++ -std=c++17 -pthread 18_traffic_light.cpp -o traffic_light
 */

#include <iostream>
#include <thread>
#include <mutex>
#include <vector>
#include <random>
#include <functional>

enum class Road { A = 1, B = 2 };

class TrafficLight {
    /*
     * EXPLANATION:
     * Only one road has green at a time.
     * Change light only when car from other road arrives.
     * Minimize unnecessary light changes.
     */
private:
    Road current_green_ = Road::A;
    std::mutex mutex_;

public:
    void carArrived(int carId, Road road,
                    std::function<void()> turnGreen,
                    std::function<void()> crossCar) {
        std::lock_guard<std::mutex> lock(mutex_);
        
        if (current_green_ != road) {
            turnGreen();
            current_green_ = road;
        }
        
        crossCar();
    }
};

class TrafficLightFair {
    /*
     * EXPLANATION:
     * Fair version: Track consecutive cars per road.
     * Switch after N cars to prevent starvation.
     */
private:
    Road current_green_ = Road::A;
    int consecutive_count_ = 0;
    const int max_consecutive_ = 5;
    std::mutex mutex_;
    std::condition_variable cv_;
    int waiting_a_ = 0;
    int waiting_b_ = 0;

public:
    void carArrived(int carId, Road road,
                    std::function<void()> turnGreen,
                    std::function<void()> crossCar) {
        std::unique_lock<std::mutex> lock(mutex_);
        
        if (road == Road::A) ++waiting_a_;
        else ++waiting_b_;
        
        // Wait for green or forced switch
        cv_.wait(lock, [&]() {
            if (current_green_ == road) return true;
            
            // Force switch for fairness
            Road other = (road == Road::A) ? Road::B : Road::A;
            int other_waiting = (road == Road::A) ? waiting_b_ : waiting_a_;
            
            if (consecutive_count_ >= max_consecutive_ && other_waiting == 0) {
                current_green_ = road;
                consecutive_count_ = 0;
                turnGreen();
                return true;
            }
            return false;
        });
        
        if (road == Road::A) --waiting_a_;
        else --waiting_b_;
        
        ++consecutive_count_;
        crossCar();
        
        cv_.notify_all();
    }
};

int main() {
    TrafficLight light;
    std::mutex cout_mutex;
    
    auto car = [&](int carId, Road road) {
        std::this_thread::sleep_for(std::chrono::milliseconds(rand() % 100));
        
        auto turnGreen = [&]() {
            std::lock_guard<std::mutex> lock(cout_mutex);
            std::cout << "Light changed to Road " 
                      << (road == Road::A ? "A" : "B") << std::endl;
        };
        
        auto crossCar = [&]() {
            std::lock_guard<std::mutex> lock(cout_mutex);
            std::cout << "Car " << carId << " from Road " 
                      << (road == Road::A ? "A" : "B") << " crossed" << std::endl;
        };
        
        light.carArrived(carId, road, turnGreen, crossCar);
    };
    
    std::vector<std::thread> threads;
    
    // Cars from Road A
    for (int i = 0; i < 5; ++i) {
        threads.emplace_back(car, i, Road::A);
    }
    
    // Cars from Road B
    for (int i = 5; i < 10; ++i) {
        threads.emplace_back(car, i, Road::B);
    }
    
    for (auto& t : threads) t.join();
    
    std::cout << "All cars passed!" << std::endl;
    return 0;
}

File: threading/python/18_traffic_light.py
"""
QUESTION 18: Traffic Light Controlled Intersection (LeetCode 1279)
==================================================================

Problem: Cars from two roads approach intersection. Control traffic light
to ensure cars from only one road pass at a time.

Key Concepts: Mutual exclusion, State machines, Thread coordination
"""

import threading
import time
import random
from enum import Enum


class Direction(Enum):
    ROAD_A = 1  # North-South
    ROAD_B = 2  # East-West


class TrafficLight:
    """
    Traffic Light Controller.
    
    EXPLANATION:
    Only one road has green light at a time.
    Cars must wait if their road has red light.
    Change light only when car from other road arrives.
    
    Key Design:
    - Track current green road
    - Lock ensures atomic light checks/changes
    - Don't change light unnecessarily (reduce context switches)
    """
    
    def __init__(self):
        self.current_green = Direction.ROAD_A
        self._lock = threading.Lock()
    
    def car_arrived(self, car_id: int, road: Direction, direction: str,
                    turn_green: Callable, cross: Callable):
        """
        Called when car arrives at intersection.
        
        Args:
            car_id: Unique car identifier
            road: Which road the car is on (A or B)
            direction: Where car is going (for printing)
            turn_green: Call to turn light green
            cross: Call when car crosses
        """
        with self._lock:
            if self.current_green != road:
                # Need to change light
                turn_green()
                self.current_green = road
            
            # Now safe to cross
            cross()


class TrafficLightFair:
    """
    Fair Traffic Light (Prevents Starvation).
    
    EXPLANATION:
    Track waiting cars on each road.
    After N cars pass, check if other road has waiting cars.
    If so, switch light to prevent starvation.
    """
    
    def __init__(self, max_consecutive: int = 5):
        self.current_green = Direction.ROAD_A
        self.consecutive_count = 0
        self.max_consecutive = max_consecutive
        self.waiting = {Direction.ROAD_A: 0, Direction.ROAD_B: 0}
        self._lock = threading.Lock()
        self._condition = threading.Condition(self._lock)
    
    def car_arrived(self, car_id: int, road: Direction,
                    turn_green: Callable, cross: Callable):
        with self._condition:
            self.waiting[road] += 1
            
            # Wait for our turn
            while self.current_green != road:
                # Check if we should switch
                other = Direction.ROAD_B if road == Direction.ROAD_A else Direction.ROAD_A
                if (self.consecutive_count >= self.max_consecutive and 
                    self.waiting[road] > 0):
                    # Force switch
                    turn_green()
                    self.current_green = road
                    self.consecutive_count = 0
                    break
                
                self._condition.wait()
            
            # Cross
            self.waiting[road] -= 1
            self.consecutive_count += 1
            cross()
            
            # Check if should switch for fairness
            other = Direction.ROAD_B if road == Direction.ROAD_A else Direction.ROAD_A
            if (self.consecutive_count >= self.max_consecutive and 
                self.waiting[other] > 0 and self.waiting[road] == 0):
                self.current_green = other
                self.consecutive_count = 0
                self._condition.notify_all()


def simulate():
    light = TrafficLight()
    
    def car(car_id: int, road: Direction, direction: str):
        time.sleep(random.uniform(0, 0.5))  # Random arrival
        
        def turn_green():
            print(f"Light changed to {road.name}")
        
        def cross():
            print(f"Car {car_id} from {road.name} going {direction}")
        
        light.car_arrived(car_id, road, direction, turn_green, cross)
    
    threads = []
    
    # Cars from Road A
    for i in range(5):
        t = threading.Thread(target=car, args=(i, Direction.ROAD_A, "North"))
        threads.append(t)
    
    # Cars from Road B
    for i in range(5, 10):
        t = threading.Thread(target=car, args=(i, Direction.ROAD_B, "East"))
        threads.append(t)
    
    random.shuffle(threads)
    
    for t in threads:
        t.start()
    for t in threads:
        t.join()
    
    print("All cars passed!")


if __name__ == "__main__":
    simulate()

